### ISSUE TITLE
compiler not restricted types of hash keys

### ISSUE THREAD

**Issue by kostya** [crystal-lang/crystal#1764]
``` ruby
x = {1 => 2}
p x["bla"]?
```

expected compile error.
in my code i trying to find where the bug, until i find that i pass wrong object to hash[]?

**Comment by asterite** [comment]
This is related to #988 . We might want to do add type restrictions there.

**Comment by asterite** [comment]
A relevant question is, does `[1, 2, 3] - [1, 'a', 3]` is valid code? When I added these type restrictions some Set specs stopped compiling, specifically some that do `Set{1, 2, 3} - Set{1, 'a', 3}`, because Set uses Hash internally and then it becomes invalid to delete an `Int32 | Char` from a Hash whose key is Int32.

This definitely needs a lot of thought.

**Comment by jhass** [comment]
I'd say it's valid, since the operation uses element equality. While `[1, 2] - ['a', 'b']` seems odd on first glance, you'd want `[1, 2] - [2.0, 3.0]` or even just `[1u8, 2u8] - [2i32, 3i32]` to work.

**Comment by kostya** [comment]
what if add this restriction by compile key? this restriction can help to find many bugs in programs.

**Comment by jhass** [comment]
That's similar to warnings. If the compiler doesn't enforce something, somebody will depend on the standard behaviour and write a library that gets popular, everybody uses that library and can't use the flag, nobody uses the flag and it becomes redundant.

**Comment by kostya** [comment]
i also spend some time to find bug, from ported ruby code.

minimized example:
```ruby
valid_wday = Set{1, 3}
t = Time.new 2015, 10, 1

result = [] of Int32
while t.month == 10
  result << t.day if valid_wday.includes?(t.day_of_week)
  t += 1.day
end

p result
```

result is empty, because `t.day_of_week` not a Int32
to fix it, just `t.day_of_week.to_i`

but i think compiler can help to find this bug, by key restriction

**Comment by asterite** [comment]
Yes, definitely. And I'll admit I also had some bugs because of wrong types to delete. And other algorithms that work on any type (like `Set#-`) can be written in a different way once we put these restrictions.

/cc @waj

**Comment by kostya** [comment]
i think in case of `[1, 2, 3] - [1, 'a', 3]` should be used special methods without checking types (which user can use when he understand what he doing), but by default it should check type of key for set, hash and array in `includes?`, `delete` and other methods

**Comment by miketheman** [comment]
#triage 
This appears to be fixed by #1851 - but that has since become un-mergable.
A modern repro of the minimized code in the most recent comment shows the problem persists: https://carc.in/#/r/11pb

**Comment by asterite** [comment]
Closing as duplicate of #988 because it's pretty much the same thing with a different method
